cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java

```diff
@@ -202,4 +202,72 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
     }
   }
 
+  public void restartServiceInstances(HttpRequest request, HttpResponder responder, final String serviceName) {
+    long startTimeMs = System.currentTimeMillis();
+    boolean isSuccess = true;
+    try {
+      if (!serviceManagementMap.containsKey(serviceName)) {
+        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
+        return;
+      }
+
+      // Need to get Twill controller for the service.
+      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+      if (masterServiceManager.restartAllInstances()) {
+        responder.sendStatus(HttpResponseStatus.OK);
+      } else {
+        isSuccess = false;
+        responder.sendString(HttpResponseStatus.BAD_REQUEST, "Operation did not succeed");
+      }
+
+    } catch (Exception e) {
+      isSuccess = false;
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                           String.format("Error starting instances for service: %s", serviceName));
+    } finally {
+      // Record the time
+      long endTimeMs = System.currentTimeMillis();
+      serviceStore.setRestartInstancesRequest(serviceName, startTimeMs, endTimeMs, isSuccess, null);
+    }
+  }
+
+  public void restartServiceInstance(HttpRequest request, HttpResponder responder, final String serviceName,
+                                     final int instanceId) {
```

Comment:
I rename the other as restartAllServiceInstances to separate restart all or an instance for a system.